### PR TITLE
fix(clear): bidirectional rate tracking and proper CloneState

### PR DESCRIPTION
**Pull request by FranckDSA**
## Summary

Fixes issues in the Clear protocol integration:

- **Bidirectional rate tracking**: Clear uses oracle-based pricing which is not symmetric. The tracker now fetches rates for ALL directions (i→j AND j→i) instead of only i<j pairs
- **Proper CloneState**: Deep copy of nested `Reserves` map to avoid race conditions during parallel route simulations
- **No fallback pricing**: Returns 0 when no rate is available instead of assuming 1:1 (which doesn't make sense for a depeg arbitrage protocol)

___

### **PR Type**
Bug fix

___

### **Description**
- Implement bidirectional rate tracking for oracle-based pricing
  - Tracker now tests all swap directions (i→j AND j→i) instead of only i<j pairs
  - Clear protocol uses non-symmetric oracle pricing, requiring both directions

- Fix CloneState to properly deep copy nested Reserves map
  - Prevents race conditions during parallel route simulations
  - Creates independent instances of all nested structures

- Simplify estimateAmountOut to use stored directional rates
  - Removes index normalization logic that assumed symmetric pricing
  - Returns 0 when rate unavailable instead of assuming 1:1 conversion

- Expand test coverage with bidirectional rate scenarios
  - Tests depeg scenarios (1:0.95 rate asymmetry)
  - Validates CloneState independence and error handling

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Pool Tracker"] -->|"Test all directions i→j AND j→i"| B["Bidirectional Rates"]
  B -->|"Store in map[i][j]"| C["Extra.Reserves"]
  C -->|"Direct lookup tokenIn→tokenOut"| D["estimateAmountOut"]
  D -->|"Calculate output"| E["CalcAmountOut"]
  F["CloneState"] -->|"Deep copy"| C
  F -->|"Create new RWMutex"| G["Independent Clone"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pool_tracker.go</strong><dd><code>Implement bidirectional rate tracking for all swap directions</code></dd></summary>
<hr>

pkg/liquidity-source/clear/pool_tracker.go

<ul><li>Changed loop structure to test all token pairs in both directions (i→j <br>AND j→i)<br> <li> Removed i<j constraint that only tested unidirectional pairs<br> <li> Added continue statement to skip same-token pairs (i==j)<br> <li> Updated comments to clarify oracle pricing is non-symmetric</ul>

</details>

  </td>
  <td><a href="https://github.com/KyberNetwork/kyberswap-dex-lib/pull/1237/files#diff-cb8a45f316e7b08cf0e17bbec5bf775c32ea61a68abca82a45fc56e0a20ceec2">+8/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>pool_simulator.go</strong><dd><code>Simplify rate lookup and implement proper deep cloning</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/liquidity-source/clear/pool_simulator.go

<ul><li>Simplified estimateAmountOut to directly use <br>tokenInIndex→tokenOutIndex lookup without index normalization<br> <li> Removed logic that swapped indices for i>j cases, now stores both <br>directions explicitly<br> <li> Updated rate variable to directly access PreviewSwapResult from map<br> <li> Implemented proper CloneState with deep copy of Info.Reserves and <br>nested extra.Reserves map<br> <li> Added RWMutex initialization and independent cloning of all big.Int <br>values<br> <li> Added comment to UpdateBalance explaining oracle-based pricing doesn't <br>require state updates</ul>

</details>

  </td>
  <td><a href="https://github.com/KyberNetwork/kyberswap-dex-lib/pull/1237/files#diff-91a221a252ae94a9c45e88857805b102b95b2cc0ca78e28b821f79894f9b5777">+55/-15</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pool_simulator_test.go</strong><dd><code>Rewrite tests for bidirectional rate format and edge cases</code></dd></summary>
<hr>

pkg/liquidity-source/clear/pool_simulator_test.go

<ul><li>Replaced old test helper with createTestPool that accepts <br>bidirectional rates<br> <li> Rewrote CalcAmountOut tests to use bidirectional rate format with <br>separate 0→1 and 1→0 rates<br> <li> Added test case demonstrating depeg scenario (1:0.95 rate asymmetry)<br> <li> Added new test for zero rates and missing rates returning 0<br> <li> Added CloneState test verifying deep copy independence<br> <li> Added tests for invalid token and invalid amount input validation</ul>

</details>

  </td>
  <td><a href="https://github.com/KyberNetwork/kyberswap-dex-lib/pull/1237/files#diff-0c8ab5a32b46f84a4dd4bee6f4e4f4b6737e59f7eaecd05b9a50cc9fc0e40ad4">+219/-29</a></td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: pkg/liquidity-source/clear/pool_simulator.go [hunk 1]
```diff
@@ -190,10 +189,51 @@ func (p *PoolSimulator) CalcAmountOutWithRPC(
...
-	return p
+	cloned := &PoolSimulator{
+		RWMutex:     &sync.RWMutex{},
+		Pool:        p.Pool,
+		staticExtra: p.staticExtra,
+		gas:         p.gas,
+	}
+
+	// Deep copy reserves
+	cloned.Info.Reserves = make([]*big.Int, len(p.Info.Reserves))
```
Comment (NgoKimPhu) [inline pkg/liquidity-source/clear/pool_simulator.go]: It should be safe for concurrent use, why do we need deep clone here?
